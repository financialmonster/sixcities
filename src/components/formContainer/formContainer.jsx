import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {bindActionCreators} from 'redux';

import {fetchReview, resetReviewError} from '../../actions/comments.js';
import Spinner from '../spinner/spinner.jsx';
import {hasUserComment} from '../../selectors.js';
import Form from '../form/form.jsx';

const FormContainer = ({isLoggedIn, fetchReview, id, isReviewLoading, isReviewError, hasUserComment,
    resetReviewError}
) => {
    useEffect(() => {        
        resetReviewError();
    }, [id, resetReviewError]);

    if(!isLoggedIn || hasUserComment) {        
        return null;
    }

    if(isReviewLoading) {
        return <Spinner />;
    }

    const formSubmitHandler = (evt, rating, comment) => {
        evt.preventDefault();

        const userData = {
            rating,
            comment
        }
        
        fetchReview(userData, id);
    }
    
    return <Form onSubmit={formSubmitHandler} />;
}

FormContainer.propTypes = {
    isLoggedIn: PropTypes.bool.isRequired,
    fetchReview: PropTypes.func.isRequired,
    id: PropTypes.number.isRequired,
    isReviewLoading: PropTypes.bool.isRequired,
    isReviewError: PropTypes.bool.isRequired,
    hasUserComment: PropTypes.bool.isRequired,
    resetReviewError: PropTypes.func.isRequired
}

const mapStateToProps = ({auth: {isLoggedIn, email}, comments:{isReviewLoading, isReviewError, comments}}) => {
    return {
        isLoggedIn,
        isReviewLoading,
        isReviewError,
        hasUserComment: hasUserComment(email, comments, isLoggedIn)
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchReview: bindActionCreators(fetchReview, dispatch),
        resetReviewError: bindActionCreators(resetReviewError, dispatch)
    }
}

export {FormContainer};
export default connect(mapStateToProps, mapDispatchToProps)(FormContainer);
