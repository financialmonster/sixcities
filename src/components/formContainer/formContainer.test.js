import React from 'react';
import renderer from 'react-test-renderer';

import {FormContainer} from './formContainer.jsx';

describe(`Correctly renders the FormContainer component`, () => {
    it(`when isLoggedIn property is equal false and hasUserReview property is equal false`, () => {
        const tree = renderer.create(
            <FormContainer fetchReview={() => {}} resetReviewError={() => {}} isReviewLoading={false}
                isReviewError={false} isLoggedIn={false} hasUserReview={false} id={2}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when hasUserReview property is equal true and isLoggedIn property is equal true`, () => {
        const tree = renderer.create(
            <FormContainer fetchReview={() => {}} resetReviewError={() => {}} isReviewLoading={false}
                isReviewError={false} isLoggedIn={true} hasUserReview={true} id={2}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isReviewLoading property is equal true, hasUserReview property is equal false and isLoggedIn property is
        equal true`,
    () => {
        const tree = renderer.create(
            <FormContainer fetchReview={() => {}} resetReviewError={() => {}} isReviewLoading={true}
                isReviewError={false} isLoggedIn={true} hasUserReview={false} id={2}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isReviewError property is equal true, hasUserReview property is equal false, isLoggedIn property is equal
        true and isReviewLoading property is equal false`,
    () => {
        const tree = renderer.create(
            <FormContainer fetchReview={() => {}} resetReviewError={() => {}} isReviewLoading={false}
                isReviewError={true} isLoggedIn={true} hasUserReview={false} id={2}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isReviewError, hasUserReview, isReviewLoading properties are equal false and isLoggedIn property is equal
        true`,
    () => {
        const tree = renderer.create(
            <FormContainer fetchReview={() => {}} resetReviewError={() => {}} isReviewLoading={false}
                isReviewError={false} isLoggedIn={true} hasUserReview={false} id={2}
            />)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
