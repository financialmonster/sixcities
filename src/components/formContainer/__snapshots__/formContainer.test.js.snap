// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders the FormContainer component when hasUserReview property is equal true and isLoggedIn property is equal true 1`] = `null`;

exports[`Correctly renders the FormContainer component when isLoggedIn property is equal false and hasUserReview property is equal false 1`] = `null`;

exports[`Correctly renders the FormContainer component when isReviewError property is equal true, hasUserReview property is equal false, isLoggedIn property is equal
        true and isReviewLoading property is equal false 1`] = `
<div
  className="cities__status-wrapper tabs__content"
>
  <b
    className="cities__status"
  >
    We're sorry
  </b>
  <p
    className="cities__status-description"
  >
    It's been an error during the process of 
    uploading of your review
    . Please, reload the page and try again
  </p>
</div>
`;

exports[`Correctly renders the FormContainer component when isReviewError, hasUserReview, isReviewLoading properties are equal false and isLoggedIn property is equal
        true 1`] = `
<form
  action="#"
  className="reviews__form form"
  method="post"
  onSubmit={[Function]}
>
  <label
    className="reviews__label form__label"
    htmlFor="review"
  >
    Your review
  </label>
  <div
    className="reviews__rating-form form__rating"
  >
    <input
      className="form__rating-input visually-hidden"
      id="5-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={5}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="5-stars"
      title="perfect"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="4-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={4}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="4-stars"
      title="good"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="3-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={3}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="3-stars"
      title="not bad"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="2-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={2}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="2-stars"
      title="badly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
    <input
      className="form__rating-input visually-hidden"
      id="1-stars"
      name="rating"
      onChange={[Function]}
      type="radio"
      value={1}
    />
    <label
      className="reviews__rating-label form__rating-label"
      htmlFor="1-stars"
      title="terribly"
    >
      <svg
        className="form__star-image"
        height="33"
        width="37"
      >
        <use
          xlinkHref="#icon-star"
        />
      </svg>
    </label>
  </div>
  <textarea
    className="reviews__textarea form__textarea"
    id="review"
    name="review"
    onChange={[Function]}
    placeholder="Tell how was your stay, what you like and what can be improved"
    value=""
  />
  <div
    className="reviews__button-wrapper"
  >
    <p
      className="reviews__help"
    >
      To submit review please make sure to set 
      <span
        className="reviews__star"
      >
        rating
      </span>
       and describe your stay with at least 
      <b
        className="reviews__text-amount"
      >
        50 characters
      </b>
      .
    </p>
    <button
      className="reviews__submit form__submit button"
      disabled={true}
      title="Submit review"
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Correctly renders the FormContainer component when isReviewLoading property is equal true, hasUserReview property is equal false and isLoggedIn property is
        equal true 1`] = `
<div
  className="spinner"
>
  <div
    className="spinnerContent"
  >
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</div>
`;
